lib/HLSL/DxilTypeSystem.cpp

```diff
@@ -253,6 +253,55 @@ StructType *DxilTypeSystem::GetNormFloatType(CompType CT, unsigned NumComps) {
   return pStructType;
 }
 
+void DxilTypeSystem::CopyTypeAnnotation(const llvm::Type *Ty,
+                                        DxilTypeSystem &src) {
+  if (isa<PointerType>(Ty))
+    Ty = Ty->getPointerElementType();
+
+  while (isa<ArrayType>(Ty))
+    Ty = Ty->getArrayElementType();
+
+  if (const StructType *ST = dyn_cast<StructType>(Ty)) {
+    // Already exist.
+    if (GetStructAnnotation(ST))
+      return;
+
+    if (DxilStructAnnotation *annot = src.GetStructAnnotation(ST)) {
+      DxilStructAnnotation *dstAnnot = AddStructAnnotation(ST);
+      // Copy the annotation.
+      *dstAnnot = *annot;
+      // Copy field type annotations.
+      for (Type *Ty : ST->elements()) {
+        CopyTypeAnnotation(Ty, src);
+      }
+    }
+  }
```

Comment:
Ok, makes sense. In that case an early check and return would make it clear.

    // Only struct type has annotation
    if (!isa<StructType>)
        return;